stumpy/core.py

```diff
@@ -2559,3 +2559,95 @@ def _select_P_ABBA_value(P_ABBA, k, custom_func=None):
             MPdist = partition[k]
 
     return MPdist
+
+
+@njit(parallel=True)
+def _merge_topk_PI(PA, PB, IA, IB):
+    """
+    Merge two top-k matrix profiles PA and PB, and update PA (in place) while
+    always choosing values of PA over values of PB in case of ties. Also, update
+    IA accordingly.
+
+    Parameters
+    ----------
+    PA : numpy.ndarray
+        A (top-k) matrix profile, with ndim of 2, where values in each row are
+        sorted in ascending order. Also, it needs to be the same shape as PB.
+
+    PB : numpy.ndarray
+        A (top-k) matrix profile, with ndim of 2, where values in each row are
+        sorted in ascending order. Also, it needs to be the same shape as PA.
+
+    IA : numpy.ndarray
+        A (top-k) matrix profile indices, corresponding to PA
+
+    IB : numpy.ndarray
+        A (top-k) matrix profile indices, corresponding to PB
+
+    Returns
+    -------
+    None
+    """
+    for i in prange(PB.shape[0]):
+        start = 0
+        stop = np.searchsorted(PA[i], PB[i, -1], side="right")
+
+        for j in range(PB.shape[1]):
+            if PB[i, j] < PA[i, -1]:
+                idx = np.searchsorted(PA[i, start:stop], PB[i, j], side="right") + start
+
+                _shift_insert_at_index(PA[i], idx, PB[i, j])
+                _shift_insert_at_index(IA[i], idx, IB[i, j])
+
+                start = idx
+                stop += 1  # because of shifting elements to the right by one
+
+
+@njit
+def _shift_insert_at_index(a, idx, v, shift=1):
+    """
+    Insert value `v` into array `a` at index `idx` (in place) and discard
```

Comment:
Consistent with the function name, I think we should discuss `shift` first. 

"Shift all elements in `a` starting at `idx` over by one. If `shift=1`, all elements in `a[idx:]` are shifted to the right by one element and the last element is discarded. If `shift=-1`, all elements in `a[idx:]` are shifted to the left by one element and the first element is discarded. In both cases, the length of `a` remains unchanged."

Reply:
> Consistent with the function name, I think we should discuss shift first.

That makes sense! I will revise the docstring. So, if `shift=-1`, we shift `a[:idx]` to left. So, if that is okay with you, I can write:

"If `shift=1`, all elements in `a[idx:]` are shifted to the right by one element and the last element is discarded. If `shift=-1`, all elements in a[:idx] are shifted to the left by one element and the first element is discarded. In both cases, the length of a remains unchanged."